ra-optimisation/rao-api/src/main/java/com/farao_community/farao/rao_api/RaoInput.java

```diff
@@ -71,57 +82,74 @@ public RaoInputBuilder withGlskProvider(GlskProvider glskProvider) {
-                raoInput.optimizedState = optimizedState;
-            }
-            if (Objects.isNull(perimeter)) {
-                raoInput.perimeter = Collections.singleton(raoInput.optimizedState);
-            } else {
-                raoInput.perimeter = perimeter;
-            }
-            raoInput.referenceProgram = Objects.isNull(referenceProgram) ? Optional.empty() : Optional.of(referenceProgram);
-            raoInput.glskProvider = Objects.isNull(glskProvider) ? Optional.empty() : Optional.of(glskProvider);
+            raoInput.crac = Optional.ofNullable(crac).orElseThrow(() -> requiredArgumentError("CRAC"));
 
+            checkBaseCracVariantId();
+            raoInput.baseCracVariantId = baseCracVariantId;
+
+            raoInput.network = Optional.ofNullable(network).orElseThrow(() -> requiredArgumentError("Network"));
+            raoInput.networkVariantId = networkVariantId != null ? networkVariantId : network.getVariantManager().getWorkingVariantId();
+            raoInput.optimizedState = optimizedState;
+            raoInput.perimeter = perimeter;
+            raoInput.referenceProgram = referenceProgram;
+            raoInput.glskProvider = glskProvider;
             return raoInput;
         }
+
+        private RaoInputException requiredArgumentError(String type) {
+            return new RaoInputException(format(REQUIRED_ARGUMENT_MESSAGE, type));
+        }
+
+        private void checkBaseCracVariantId() {
+            ResultVariantManager resultVariantManager = crac.getExtension(ResultVariantManager.class);
+            if (baseCracVariantId != null) {
+                if (resultVariantManager == null) {
+                    throw new RaoInputException("Base CRAC variant cannot be specified if CRAC does not have result variant manager.");
+                }
+                if (!resultVariantManager.getVariants().contains(baseCracVariantId)) {
+                    throw new RaoInputException(format("Base CRAC variant %s does not exist.", baseCracVariantId));
+                }
+            } else {
+                if (resultVariantManager != null && resultVariantManager.getVariants() != null) {
+                    throw new RaoInputException("Base CRAC variant has to be specified if CRAC already has got pre-optimization variant.");
+                }
+            }
+        }
     }
 
     private Crac crac;
+    private String baseCracVariantId;
     private State optimizedState;
     private Set<State> perimeter;
     private Network network;
-    private String variantId;
-    private Optional<ReferenceProgram> referenceProgram;
-    private Optional<GlskProvider> glskProvider;
+    private String networkVariantId;
+    private ReferenceProgram referenceProgram;
+    private GlskProvider glskProvider;
 
     private RaoInput() {
     }
 
-    public static RaoInputBuilder builder() {
-        return new RaoInputBuilder();
+    public static RaoInputBuilder create(Network network, Crac crac) {
```

Comment:
build instead of create ? otherwise I would expect a RaoInput in output of this method ?

Reply:
Are those methods still used ? It seems they are only in the tool and the tests. I think they could be deleted

Reply:
I will rename create -> build
the methods are shortcuts to builder().withCrac().withNetwork() .. (since we always have to add a crac and a network anyway). but we agree that they are not mandatory, we can delete them if you prefere